### ISSUE TITLE
istio 1.5 - authZ policy behavior causing issues

### ISSUE THREAD

**Issue by rshriram** [istio/istio#21975]
**Bug description**

The new authZ policy is additive in nature. Multiple authZ policies can be defined for the same workload. At runtime, pilot combines all of them in the Envoy config. This additive behavior causes some issues when the intent does not match the generated config. For example, consider this set of configs:

1. The system has 3 namespaces (httpbin-front, httpbin-middle, httpbin-vm)
2. The user intent is that only these namespaces should be able to talk to one another and other namespaces in the system (say bookinfo) should not be able to talk to these 3 namespaces.
3. In addition, the user wishes to enforce JWT based rules on the ingress gateway in the httpbin-front namespace for all internet traffic.

The user expectation is that the ingress gateway will only do JWT validation for traffic entering the gateway, while all internal services will do the istio SAN based validation to ensure that only services from the 3 namespaces can talk to one another. However, today, we end up generating a configuration on the gateway that contains both rules (only allow traffic from the 3 namespaces and only allow traffic with the JWT validation). Consequently, the gateway is unable to accept any internet traffic coming over one-way HTTPS.

**Expected behavior**

Be able to apply separate policies for the gateway in a namespace as well as default policies for the rest of the services in the namespace.

**Steps to reproduce the bug**
Istio config:

```yaml
apiVersion: security.istio.io/v1beta1
kind: RequestAuthentication
metadata:
  name: mns-httpbin-gateway-httpbin-front
  namespace: httpbin-front
spec:
  jwtRules:
  - issuer: tetrate.io
    jwks: '{"keys":[{"kid":"a2480531-e348-40ee-9280-890e215bf250","kty":"oct","alg":"HS256","k":"c2VjcmV0LWtleQ=="}]}'
  selector:
    matchLabels:
      app: gateway-httpbin-front
---
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: default
  namespace: httpbin-front
spec:
  rules:
  - from:
    - source:
        namespaces:
        - httpbin-front
        - httpbin-middle
        - httpbin-vm
---
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: mns-httpbin-gateway-httpbin-front
  namespace: httpbin-front
spec:
  rules:
  - to:
    - operation:
        paths:
        - /
        - /?*
  - to:
    - operation:
        paths:
        - /status/200
    when:
    - key: request.headers[x-app-trusted]
      values:
      - "true"
  - to:
    - operation:
        paths:
        - /status/200
    when:
    - key: request.auth.claims[group]
      values:
      - engineering
  - from:
    - source:
        requestPrincipals:
        - tetrate.io/user
    to:
    - operation:
        paths:
        - /status/200
    when:
    - key: request.auth.claims[group]
      values:
      - support
  selector:
    matchLabels:
      app: gateway-httpbin-front
---
```

And the generated envoy istio authN/rbac filters for the gateway is as follows:

```json
             {
              "name": "envoy.filters.http.jwt_authn",
              "typed_config": {
               "@type": "type.googleapis.com/envoy.config.filter.http.jwt_authn.v2alpha.JwtAuthentication",
               "providers": {
                "origins-0": {
                 "issuer": "tetrate.io",
                 "local_jwks": {
                  "inline_string": "{\"keys\":[{\"kid\":\"a2480531-e348-40ee-9280-890e215bf250\",\"kty\":\"oct\",\"alg\":\"HS256\",\"k\":\"c2VjcmV0LWtleQ==\"}]}"
                 },
                 "payload_in_metadata": "tetrate.io"
                }
               },
               "rules": [
                {
                 "match": {
                  "prefix": "/"
                 },
                 "requires": {
                  "requires_any": {
                   "requirements": [
                    {
                     "provider_name": "origins-0"
                    },
                    {
                     "allow_missing": {}
                    }
                   ]
                  }
                 }
                }
               ]
              }
             },
             {
              "name": "istio_authn",
              "typed_config": {
               "@type": "type.googleapis.com/istio.envoy.config.filter.http.authn.v2alpha1.FilterConfig",
               "policy": {
                "peers": [
                 {
                  "mtls": {}
                 }
                ],
                "peer_is_optional": true,
                "origins": [
                 {
                  "jwt": {
                   "issuer": "tetrate.io"
                  }
                 }
                ],
                "origin_is_optional": true,
                "principal_binding": "USE_ORIGIN"
               }
              }
             },
             {
              "name": "envoy.filters.http.rbac",
              "typed_config": {
               "@type": "type.googleapis.com/envoy.config.filter.http.rbac.v2.RBAC",
               "rules": {
                "policies": {
                 "ns[httpbin-front]-policy[default]-rule[0]": {
                  "permissions": [
                   {
                    "and_rules": {
                     "rules": [
                      {
                       "any": true
                      }
                     ]
                    }
                   }
                  ],
                  "principals": [
                   {
                    "and_ids": {
                     "ids": [
                      {
                       "or_ids": {
                        "ids": [
                         {
                          "metadata": {
                           "filter": "istio_authn",
                           "path": [
                            {
                             "key": "source.principal"
                            }
                           ],
                           "value": {
                            "string_match": {
                             "safe_regex": {
                              "google_re2": {},
                              "regex": ".*/ns/httpbin-front/.*"
                             }
                            }
                           }
                          }
                         },
                         {
                          "metadata": {
                           "filter": "istio_authn",
                           "path": [
                            {
                             "key": "source.principal"
                            }
                           ],
                           "value": {
                            "string_match": {
                             "safe_regex": {
                              "google_re2": {},
                              "regex": ".*/ns/httpbin-middle/.*"
                             }
                            }
                           }
                          }
                         },
                         {
                          "metadata": {
                           "filter": "istio_authn",
                           "path": [
                            {
                             "key": "source.principal"
                            }
                           ],
                           "value": {
                            "string_match": {
                             "safe_regex": {
                              "google_re2": {},
                              "regex": ".*/ns/httpbin-vm/.*"
                             }
                            }
                           }
                          }
                         }
                        ]
                       }
                      }
                     ]
                    }
                   }
                  ]
                 },
                 "ns[httpbin-front]-policy[mns-httpbin-tsb-gateway-httpbin-front]-rule[0]": {
                  "permissions": [
                   {
                    "and_rules": {
                     "rules": [
                      {
                       "or_rules": {
                        "rules": [
                         {
                          "header": {
                           "name": ":path",
                           "exact_match": "/"
                          }
                         },
                         {
                          "header": {
                           "name": ":path",
                           "prefix_match": "/?"
                          }
                         }
                        ]
                       }
                      }
                     ]
                    }
                   }
                  ],
                  "principals": [
                   {
                    "and_ids": {
                     "ids": [
                      {
                       "any": true
                      }
                     ]
                    }
                   }
                  ]
                 },
                 "ns[httpbin-front]-policy[mns-httpbin-tsb-gateway-httpbin-front]-rule[1]": {
                  "permissions": [
                   {
                    "and_rules": {
                     "rules": [
                      {
                       "or_rules": {
                        "rules": [
                         {
                          "header": {
                           "name": ":path",
                           "exact_match": "/status/200"
                          }
                         }
                        ]
                       }
                      }
                     ]
                    }
                   }
                  ],
                  "principals": [
                   {
                    "and_ids": {
                     "ids": [
                      {
                       "or_ids": {
                        "ids": [
                         {
                          "header": {
                           "name": "x-app-trusted",
                           "exact_match": "true"
                          }
                         }
                        ]
                       }
                      }
                     ]
                    }
                   }
                  ]
                 },
                 "ns[httpbin-front]-policy[mns-httpbin-tsb-gateway-httpbin-front]-rule[2]": {
                  "permissions": [
                   {
                    "and_rules": {
                     "rules": [
                      {
                       "or_rules": {
                        "rules": [
                         {
                          "header": {
                           "name": ":path",
                           "exact_match": "/status/200"
                          }
                         }
                        ]
                       }
                      }
                     ]
                    }
                   }
                  ],
                  "principals": [
                   {
                    "and_ids": {
                     "ids": [
                      {
                       "or_ids": {
                        "ids": [
                         {
                          "metadata": {
                           "filter": "istio_authn",
                           "path": [
                            {
                             "key": "request.auth.claims"
                            },
                            {
                             "key": "group"
                            }
                           ],
                           "value": {
                            "list_match": {
                             "one_of": {
                              "string_match": {
                               "exact": "engineering"
                              }
                             }
                            }
                           }
                          }
                         }
                        ]
                       }
                      }
                     ]
                    }
                   }
                  ]
                 },
                 "ns[httpbin-front]-policy[mns-httpbin-tsb-gateway-httpbin-front]-rule[3]": {
                  "permissions": [
                   {
                    "and_rules": {
                     "rules": [
                      {
                       "or_rules": {
                        "rules": [
                         {
                          "header": {
                           "name": ":path",
                           "exact_match": "/status/200"
                          }
                         }
                        ]
                       }
                      }
                     ]
                    }
                   }
                  ],
                  "principals": [
                   {
                    "and_ids": {
                     "ids": [
                      {
                       "or_ids": {
                        "ids": [
                         {
                          "metadata": {
                           "filter": "istio_authn",
                           "path": [
                            {
                             "key": "request.auth.principal"
                            }
                           ],
                           "value": {
                            "string_match": {
                             "exact": "tetrate.io/user"
                            }
                           }
                          }
                         }
                        ]
                       }
                      },
                      {
                       "or_ids": {
                        "ids": [
                         {
                          "metadata": {
                           "filter": "istio_authn",
                           "path": [
                            {
                             "key": "request.auth.claims"
                            },
                            {
                             "key": "group"
                            }
                           ],
                           "value": {
                            "list_match": {
                             "one_of": {
                              "string_match": {
                               "exact": "support"
                              }
                             }
                            }
                           }
                          }
                         }
                        ]
                       }
                      }
                     ]
                    }
                   }
                  ]
                 }
                }
               }
              }
             },
```

Notice how both the rules are in the gateway even though the intent is to simply do JWT authn/authz at the gateway and istio mtls san based authz post gateway.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
1.5
**How was Istio installed?**
N/A
**Environment where bug was observed (cloud vendor, OS, etc)**
kind

**Comment by yangminzhu** [comment]
> However, today, we end up generating a configuration on the gateway that contains both rules (only allow traffic from the 3 namespaces and only allow traffic with the JWT validation). Consequently, the gateway is unable to accept any internet traffic coming over one-way HTTPS.

This doesn't seem right based on the semantics of the authZ policy. The authZ policies are ORed together which means the gateway should accept either 1) traffic from the 3 namespaces OR 2) traffic with the JWT. @rshriram could you double check this behavior in your environment?

On the other hand, I think you are expecting the authZ policy to have sidecar-style overrides instead of namespace inheritance. However the namespace inheritance is not anything new in 1.5 in Authz. The AuthZ as well as the old RBAC policy has been using the namespace inheritance since the very beginning in istio 0.8.

In other words, this is the intended behavior. To achieve your goal, you need to change the following policy to use `workloadSelector` to restrict the policy to only apply to the sidecars and exclude the gateway.

```yaml
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: default
  namespace: httpbin-front
spec:
  rules:
  - from:
    - source:
        namespaces:
        - httpbin-front
        - httpbin-middle
        - httpbin-vm
```

**Comment by yangminzhu** [comment]
@rshriram Please let me know if the above response answer your question, this is actually the intended behavior of the authZ policy. I will close the issue if there is no other questions, thank you.

**Comment by rshriram** [comment]
@yangminzhu you were right. in that the policies in envoy are OR-ed and not AND-ed. Thanks for pointing that out! As to whether we need an exclusion model for the policies, I certainly think that we need a `excludeWorkloads` section in the authZ policy so that only policies specific to the API gateways can be applied and not all internal policies.

**Comment by yangminzhu** [comment]
Thanks for the confirm, closing the issue as it's the intended behavior. For `excludeWorkloads`, that would be a feature request that also affects other policies using the selector (e.g. AuthN policy), feel free to open an issue for it.
